src/sagas/vmChanges.js

```diff
@@ -177,54 +162,132 @@ function* composeProvisionSourceTemplate ({ vm, basic, disks }) {
+    )
 
-      // make sure disk format (and therefore sparse) matches the template's storageAllocation
-      if (templateStorageAllocation === 'thin') {
-        changesToTemplateDisk.format = 'cow'
-      } else if (templateStorageAllocation === 'clone') {
-        changesToTemplateDisk.format = vmStorageAllocation === 'thin' ? 'cow' : 'raw'
-      }
+    return Object.keys(changes).length === 0
+      ? false
+      : { id: disk.id, ...changes }
+  })
 
-      // did the storage domain change?
-      if (disk.storageDomainId !== templateDisk.get('storageDomainId')) {
-        diskRequiresClone = true
-        changesToTemplateDisk.storage_domains = {
-          storage_domain: [{ id: disk.storageDomainId }],
-        }
-      }
+  /*
+   * If we change ANY disk on the template, we need to send in ALL disks.
+   */
+  if (diskChanges.some(Boolean)) {
+    vmUpdates.disk_attachments = {
+      disk_attachment: diskChanges.map(
+        (changed, index) => ({
+          disk: changed || { id: disksFromTemplate[index].disk.id },
+        })
+      ),
+    }
+  }
 
-      if (Object.keys(changesToTemplateDisk).length > 1) {
-        if (vmUpdates.disk_attachments) {
-          // add another disk to clone
-          vmUpdates.disk_attachments.disk_attachment.push({ disk: changesToTemplateDisk })
-        } else {
-          // initial setup a disk to clone
-          vmUpdates.disk_attachments = {
-            disk_attachment: [{
-              disk: changesToTemplateDisk,
-            }],
-          }
-        }
-      }
-    })
+  return [vmUpdates, storageAllocation === 'clone']
+}
+
+/**
+ * Select the disk's __format__ and __sparse__ based on where the disk is coming from,
+ * what changes the user has made to the VM from the template, and what the disk's
+ * target storage domain looks like.
+ *
+ * Webadmin behavior:
+ *   [Template OptimizedFor -> VM OptimizedFor -> SD changes? ~~ Storage allocation ... disk format/sparse outcome]
+ *   (* = default)
+ *
+ *   Server  -> Server*  -> No SD changes* ~~ Clone ... format stays the same, sparse stays the same
+ *   Server  -> Server*  -> SD changes ~~ Clone     ... format stays the same, sparse changes as necessary based on format + SD
+ *   Server  -> Desktop  -> No SD changes* ~~ Thin  ... format: cow, sparse: true
+ *   Server  -> Desktop  -> SD changes ~~ Clone     ... format: cow, sparse: true (Thin settings retained)
+ *   Desktop -> Desktop* -> No SD changes* ~~ Thin  ... format: cow, sparse: true
+ *   Desktop -> Desktop* -> SD changes ~~ Clone     ... format: cow, sparse: true (Thin settings retained)
+ *   Desktop -> Server   -> No SD changes* ~~ Clone ... format: cow, sparse: true (Thin settings retained)
+ *   Desktop -> Server   -> SD changes ~~ Clone     ... format: cow, sparse: true (Thin settings retained)
+ */
+function determineTemplateDiskFormatAndSparse (templateOptimizedFor, vmOptimizedFor, baseDisk, userDisk, storageAllocation, targetSD) {
```

Comment:
>   1. can we unit test according to scenarios above?

Probably.

>   2. general remark - IMHO we should try to limit the use of ImmutableJS i.e. here with plain JS objects it would be easier read the code.

I'd rather starts stripping out immutable calls in a different PR.  Keeping the access of things held in the store mostly the same across the project makes sense to me

>   3. with 6 params it makes sense to pass them via object

Why?  I don't see any difference between 6 parameters and a single object with 6 attributes.

Reply:
> I'd rather start stripping out immutable calls in a different PR. Keeping the access of things held in the store mostly the same across the project makes sense to me

I would rather convert from ImmutableJs to plain objects after we receive the data from the store. Lets keep components plain JS.  This makes logic more readable - especially that with `?.` operator the `getIn([path])` method lost most of the importance.

> > 1. with 6 params it makes sense to pass them via object
> 
> Why? I don't see any difference between 6 parameters and a single object with 6 attributes.

Just to clarify:  I mean replacing `foo(a,b,c,d)`  with `foo({a,b,c,d})`. For me the main advantage is that the order of parameters is not important.

Reply:
> > I'd rather start stripping out immutable calls in a different PR. Keeping the access of things held in the store mostly the same across the project makes sense to me
> 
> I would rather convert from ImmutableJs to plain objects after we receive the data from the store. Lets keep components plain JS. This makes logic more readable - especially that with `?.` operator the `getIn([path])` method lost most of the importance.
> 

I understand what you're saying.  I don't want to do that in this PR because I would need to touch other sagas to keep the "drop immutable and just work with normal JS objects" consistent in the file.

> > > 1. with 6 params it makes sense to pass them via object
> > 
> > 
> > Why? I don't see any difference between 6 parameters and a single object with 6 attributes.
> 
> Just to clarify: I mean replacing `foo(a,b,c,d)` with `foo({a,b,c,d})`. For me the main advantage is that the order of parameters is not important.

I'm not worried about that at all.  Your ask could apply to all functions we have with >1 argument.  Maybe in future we can find and add a lint rule and make all functions follow that rule.